### ISSUE TITLE
jarified apps throw "Please run `rake reflect` to generate the imports"

### ISSUE THREAD

**Issue by griest024** [jruby/jrubyfx#101]
Win 8.1 64-bit
JDK 8u66
JRuby 9.1.2.0
Ran all this stuff from the git bash, in case that matters

I started with a clean install of jruby 9.1.2.0 and ran:

```
$ jruby -S gem install jrubyfx
Successfully installed jrubyfx-fxmlloader-0.4.1-java
Successfully installed jrubyfx-1.1.1-java
2 gems installed
```

Then jarified demo:

```
$ jruby -S jrubyfx-jarify samples/fxml/ --main samples/fxml/Demo.rb Demo.jar
mkdir -p C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q
cp C:\Users\Peter/.jruby-jar/jruby-complete-9.1.2.0.jar   C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q/Demo.jar
cp -r samples/fxml//ComplexControl.fxml C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q/ComplexControl.fxml
cp -r samples/fxml//Demo.fxml C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q/Demo.fxml
cp -r samples/fxml//Demo.rb C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q/Demo.rb
cp -r samples/fxml//complex_control.rb C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q/complex_control.rb
cp samples/fxml/Demo.rb C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q/jar-bootstrap.rb
cp -r C:/jruby-9.1.2.0/lib/ruby/gems/shared/gems/jrubyfx-1.1.1-java/lib/jrubyfx C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q
cp -r C:/jruby-9.1.2.0/lib/ruby/gems/shared/gems/jrubyfx-1.1.1-java/lib/jrubyfx.rb C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q
cp -r C:/jruby-9.1.2.0/lib/ruby/gems/shared/gems/jrubyfx-1.1.1-java/lib/jrubyfx_tasks.rb C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q
cp -r C:/jruby-9.1.2.0/lib/ruby/gems/shared/gems/jrubyfx-fxmlloader-0.4.1-java/lib/FXMLLoader-j8.jar C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q
cp -r C:/jruby-9.1.2.0/lib/ruby/gems/shared/gems/jrubyfx-fxmlloader-0.4.1-java/lib/fxmlloader C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q
cp -r C:/jruby-9.1.2.0/lib/ruby/gems/shared/gems/jrubyfx-fxmlloader-0.4.1-java/lib/jrubyfx-fxmlloader.rb C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q
cd C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q
jar ufe 'Demo.jar' org.jruby.JarBootstrapMain *
chmod 775 Demo.jar
cd C:/Users/Peter/jrubyfx
mv C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q/Demo.jar Demo.jar
rm -rf C:/Users/Peter/AppData/Local/Temp/jrubyfx20160529-11204-1n4it8q
```

Doesn't work when the jar is run:

```
$ java -jar Demo.jar
Please run `rake reflect` to generate the imports
```

The demo sample runs just fine from source:

```
$ jruby samples/fxml/Demo.rb
C:/jruby-9.1.2.0/lib/ruby/gems/shared/gems/jrubyfx-fxmlloader-0.4.1-java/lib/fxmlloader/elts.rb:158: warning: `<<' after local variable or literal is interpreted as binary operator
C:/jruby-9.1.2.0/lib/ruby/gems/shared/gems/jrubyfx-fxmlloader-0.4.1-java/lib/fxmlloader/elts.rb:158: warning: even though it seems like here document
saved C:/Users/Peter/jrubyfx/samples/fxml/.jrubyfx_cache/d9389de9763b047eafd2e02636bbed06906dee0d.rb to cache
FXML loaded
Send Stuff To initialized
Find by Ruby magic
AnchorPane[id=rootAnchorPane, styleClass=root]
Pane[id=root]

Find single node CSS lookup
AnchorPane[id=rootAnchorPane, styleClass=root]
Pane[id=root]

find returns nil when no match.
find! raises exceptions

Find by basic CSS selector:
        rootAnchorPane => []
        Pane => [#<Java::JavafxSceneLayout::Pane:0x406e234e>]
        #root => [#<Java::JavafxSceneLayout::Pane:0x406e234e>]
        MenuBar => [#<Java::JavafxSceneControl::MenuBar:0x221527b9>]
        Button => [#<Java::JavafxSceneControl::Button:0x9227066>, #        <Java::JavafxSceneControl::Button:0x4e2b5b74>, #<Java::JavafxSceneControl::Button:0x74358897>, #<Java::JavafxSceneControl::Button:0x2b72606e>, #<Java::JavafxSceneControl::Button:0x41e86679>, #<Java::JavafxSceneControl::Button:0x30f0d336>]

Complex CSS doesn't work:
        MenuBar Menu => []
        MenuBar MenuItem => []
        MenuItem => []
        Menu => []
        [textFill="WHITE"] => []
        [fx|id] => []
saved C:/Users/Peter/jrubyfx/samples/fxml/.jrubyfx_cache/ce9523b56dc118e65f030aa195ea33f48b8ec1ee.rb to cache
Text box contains: Hello JRubyFX
Clicked Green
Clicked Blue, Orange, or Black
Clicked Red
javafx.event.ActionEvent[source=Button@4e2b5b74[styleClass=button]'Red']
Clicked Red
javafx.event.ActionEvent[source=Button@4e2b5b74[styleClass=button]'Red']
Clicked Blue, Orange, or Black
Clicked Purple
```

**Comment by griest024** [comment]
@byteit101 I think this might be your arena.

**Comment by oreoshake** [comment]
This is only an issue on jruby 9k for me. I had to disable the sections that check if reflection has occurred:

https://github.com/BrakemanSecurityInc/jrubyfx/commit/d988324b9d5bf3b2c4bad00553d6129576de00fd and https://github.com/BrakemanSecurityInc/jrubyfx/commit/ed2279668d62c63ddba084828a8bda1fb8f26f3a.

Things ran just fine, so it's the check that is busted. Our fork is on rubygems if you want to use it in the short term, but something about `File.size? "#{File.dirname(__FILE__)}/jrubyfx/imports.rb"` isn't working in jars.

Also, I found the `in_jar?` method to be broken on 9k as well. I ended up creating an `ENV` var that is manually set: https://github.com/BrakemanSecurityInc/jrubyfx/commit/5e36e097a33ed275741342d92d91b36628c7c68b

These are all terrible hacks around potential issues, but "it works"

**Comment by griest024** [comment]
Thanks this will be helpful. I'll let you know if I come up with a less hacky solution. Chances are its a jrubyfx bug, even though its only in 9k. I might make a PR.

**Comment by enebo** [comment]
@oreoshake Could you insert a line to print out what the expanded name is for ```File.size? "#{File.dirname(__FILE__)}/jrubyfx/imports.rb"``` ?  Also I am curious is another File method like exist? will work with the path.  If it doesn't then it is likely a jar handling bug in JRuby itself (which I suspect only manifests on windows).

**Comment by oreoshake** [comment]
@enebo:

```
uri:classloader:/vendor/bundle/jruby/2.3.0/gems/jrubyfx-master-1.1.1.brakemanpro6-java/lib/jrubyfx/imports.rb
```

**Comment by enebo** [comment]
Just noticed original example is also broken on MacOS.  @mkristian anything jump out here?

**Comment by mkristian** [comment]
@enebo  the path looks ok though I did not see where `vendor/bundle` as I could not see any bundler involved in the examples. `File.size?` is something else as it means that we have a way to detemine the size of file. if the `classloader.getResource` gives an jar or file url then this could be used but in all other cases you just can get an `InputStream` to the resource, i.e. have to read the file to get its size. but I do not remember any such implementation with classloader resource

**Comment by oreoshake** [comment]
The `/vendor/bundle` may be specific to my app. I've created a test app at https://github.com/oreoshake/jrubyfx-9k-jar-problems that runs into this exact issue.

**Comment by oreoshake** [comment]
I should have mentioned that you must be running jruby head to run into the error ⬆️

**Comment by enebo** [comment]
@mkristian I used original reporters script to try this.  If I change from size? to exist? in the two places in this codebase then I make it past that problem.  What is weird is this works on 1.7 (or I presume it does).

**Comment by mkristian** [comment]
@enebo yes, `exists?` works but not `size?` as it is not implemented: https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/util/URLResource.java#L100

with 1.7.x it works as it most probably uses the `JarResource`. the `URLResource` can easily produce the length for file or jar URLs but for others it has to read in the complete file to determine its size. but with 9k we basically use `uri:classloader:` unless the application uses those `jar:file:my.jar!/my/ruby.rb` explicitly.

**Comment by enebo** [comment]
So I implemented size? for uri:classloader: resources and this will be in jruby 9.1.3.0.  There is still an issue though.  File.directory? is returning true on a classloader resource as it is a directory but that directory cannot then be mkdir'd since it is read only.  I am working on a fix to fxloader.  I cannot really say JRuby is wrong even though behavior is a little different.  I will open a PR on that soon.

**Comment by enebo** [comment]
In reference to my last comment this last problem ends up being a bigger issue in that 1.7 and 9k both prefix base directory in different ways jar: url:classloader.  I will need to dig deeper to figure out if 9k has an issue or if we need to deal with both ways.

**Comment by griest024** [comment]
Keep in mind that [this commit](https://github.com/jruby/jrubyfx/pull/107/commits/4a3ea007be501cd51aa9e5cdfefcc7fdd7408c74) will probably affect this issue.

**Comment by byteit101** [comment]
JRubyFX 2.0 has been released. Your issue may have been fixed in 2.0. If your issue still applies to JRubyFX 2.0, please reopen this issue.
